TEP-0045: When Expressions in Finally Tasks

jerop opened the pull request:
Users can guard execution of `Tasks` using `WhenExpressions`,
but that is not supported in `Finally Tasks`.

This TEP describes the need for supporting `WhenExpressions` in
`Finally Tasks` not only to provide efficient guarded execution
but also to improve the reusability of `Tasks`.

Given we've recently added support for `Results` and `Status` in
`Finally Tasks`, this is an opportune time to enable `WhenExpressions`
in `Finally Tasks`.

/kind tep
/cc @pritidesai @bobcatfish @sbwsg

bobcatfish commented:
/assign @pritidesai 
/assign @sbwsg

pritidesai commented on teps/0045-guard-finally-tasks-with-whenexpressions.md:
```diff
@@ -0,0 +1,303 @@
+
+Users can guard execution of `Tasks` using `WhenExpressions`, but that is not supported in `Finally Tasks`. This TEP 
+describes the need for supporting `WhenExpressions` in `Finally Tasks` not only to provide efficient guarded execution
+but also to improve the reusability of `Tasks`. Given we've recently added support for `Results` and `Status` in 
```
I would recommend reflecting this in the title as well - its not just about guarding, its more about reusability of a task in the finally section.

jerop replied:
Great idea! Thanks :)

Updated it to `Reusability and Guarded Execution of Finally Tasks`

pritidesai commented on teps/0045-guard-finally-tasks-with-whenexpressions.md:
```diff
@@ -0,0 +1,303 @@
+-->
+
+The primary goal of this proposal is to allow guarding execution of `Finally Tasks` using `WhenExpressions`. A secondary
+goal is to improve the [reusability](https://github.com/tektoncd/community/blob/master/design-principles.md#reusability)
```
I would switch this and make reusability as the primary goal.

jerop replied:
made them "co-goals", but mentioned reusability first

pritidesai commented on teps/0045-guard-finally-tasks-with-whenexpressions.md:
```diff
@@ -0,0 +1,303 @@
+          - input: failure
+            operator: in
+            values:
+              - $(tasks.build.status)
```
This looks little weird. can we change the when expression to switch input with values?

```
        when:
          - input: $(tasks.build.status)
            operator: in
            values: ["Failed"]
```

jerop replied:
done, thanks :)

sbwsg reviewed (approved):
Adding this feature makes sense to me and composes well with the existing access Finally tasks get to results and status.  👍

pritidesai commented:
thanks @jerop 

This looks good in proposed state. Please add more design details while making it implementable. 

`When` expressions are so far evaluated after successful or skipped tasks. Will need to address failure scenario in the when expression evaluation i.e. a result reference in when expression could be missing and the finally task ends up in the list of skipped tasks v/s when expression with mismatching `inputs` and `values` evaluating to false. This will be no different than how we have implemented task results in `finally`. 

This also makes me think `guard` and `finally` together sounds very odd. Should we change the title again to drop `guarding`? `when` expression is bringing reusability rather than guarding it compared to non-finally tasks.

jerop commented:
thanks for the reviews @sbwsg & @pritidesai!! 🙏🏾 

> This looks good in proposed state. Please add more design details while making it implementable.
 
I've added more details about using `Parameters`, `Results` and `Execution Status`, and also addressed the validation and failure scenarios -- please let me know if there's something more you'd like to  see

> `When` expressions are so far evaluated after successful or skipped tasks. Will need to address failure scenario in the when expression evaluation i.e. a result reference in when expression could be missing and the finally task ends up in the list of skipped tasks v/s when expression with mismatching `inputs` and `values` evaluating to false. This will be no different than how we have implemented task results in `finally`.

Yes, in both scenarios, they'd be skipped and added to the `Skipped Tasks` in status -- another reason to look into adding `Reason` to `Skipped Tasks` and revisit the general `Status` design

https://github.com/tektoncd/community/blob/bf680b76ff10b52b541dd05a9277a37e5dcbc39a/teps/0045-reusability-and-guarding-of-tasks-finally.md#L232-L234

> This also makes me think `guard` and `finally` together sounds very odd. Should we change the title again to drop `guarding`? `when` expression is bringing reusability rather than guarding it compared to non-finally tasks.

@pritidesai, what do you think of `Reusability and Guarding of Tasks in Finally` for the title?

I'd say the same argument about `WhenExpressions` improving reusability of Tasks in `Finally` also applies to the `DAG` (non-finally), plus I think excluding _guarding_ in the title would imply this TEP is focused more broadly on the _reusability_ of `Finally Tasks` beyond supporting `WhenExpressions`

pritidesai commented:
thanks @jerop, we could add more relevant examples for using [task results](https://github.com/tektoncd/community/pull/317/files#diff-8f1f065c6c4af5eadd0c13ab0b7ebc4d9e5c13a00c4c3228bcdb5945e7e6f62aR209) and [params](https://github.com/tektoncd/community/pull/317/files#diff-8f1f065c6c4af5eadd0c13ab0b7ebc4d9e5c13a00c4c3228bcdb5945e7e6f62aR243) with when expressions but its not blocking this proposal.

Thanks for all the updates 🙏 

/lgtm
